Use buildx `docker` driver, add `--cache-to type=inline`

ibeckermayer opened the pull request:
Based on pain experienced attempting to update the `Dockerfile-centos7-fips` image. See the commit messages for details.

Do not merge until: 
- [x] https://github.com/gravitational/teleport.e/pull/3300 is merged and e-ref is updated.
- [x] `Build CentOS 7 Buildbox Assets Images` succeeds, which updates the `build-centos7-assets` image with the new `BUILDKIT_INLINE_CACHE=1` cache: [run](https://github.com/gravitational/teleport/actions/runs/7717269335), [result](https://github.com/gravitational/teleport/pkgs/container/teleport-buildbox-centos7-assets/172883965?tag=teleport16-amd64)
- [x] In the same vain as the prev bullet point, `make -C build.assets/ build-centos7-assets` runs on my arm64 dev machine and the result is manually pushed (we don't have a workflow for this yet): [result](https://github.com/gravitational/teleport/pkgs/container/teleport-buildbox-centos7-assets/172885262?tag=teleport16-arm64)
- [x] ~A `Build CI Buildbox Images` succeeds (TODO): https://github.com/gravitational/teleport/actions/workflows/build-ci-buildbox-images.yaml~: This can't be started manually.
- [x] A dev build succeeds: https://drone.platform.teleport.sh/gravitational/teleport/33501

Commit:
Use buildx's `docker` driver for buildbox-oriented GHA jobs

setup-buildx-action uses the docker-container buildx driver by default,
which is useful for advanced buildx features like customizing buildkit,
doing multi-arch builds, and advanced cache import/export (particularly
as it pertains to multi-arch builds).

However, we're not using these features in our buildbox-oriented GHA
jobs, and in fact the `docker-container` driver is not well suited for
our `buildbox-centos7-fips` Make target, which relies on "old school"
(`docker` driver) behavior that expects to find an updated version of
the image built by `buildbox-centos7` (`Dockerfile-centos7`) to pass as
the FROM argument to the `buildbox-centos7-fips`.
(`Dockerfile-centos7-fips`). Since the `docker-container` driver runs
each build in a separate container, the `buildbox-centos7-fips` build
can't see the updated image built by `buildbox-centos7`, and so resorts
to the stale image in the remote docker repository.

We could restructure our docker buildx commands to work around this, but
that would result in a much more complicated, distributed build process
that would potentially increase cloud spend, for marginal benefit to the
status quo at best. Instead we can just use the `docker` driver for our
buildbox-oriented GHA to keep things simple, and only update to the
`docker-container` driver when we need to use its advanced features.

Refs:
- https://github.com/docker/setup-buildx-action
- https://docs.docker.com/build/drivers/

Commit:
Adds `BUILDKIT_INLINE_CACHE=1` to `buildx build` commands that use `--cache-from`.
The `--cache-from` flag is used to specify a cache source. This "cache" is
a special buildx cache that improves on the traditional docker layer-based cache.
While there are advanced options for exporting/importing this advanced cache separately
from the image (layer-based cache), the simplest way to make use of it is to just
embed the advanced cache inline in the image itself. This is what this BUILDKIT_INLINE_CACHE=1
does.

Commit:
Release v16.0.0-dev.isaiah.18

camscale commented:
I think instead of adding a parameter to the CI docker driver setup, we should just add `--builder default` in the `Makefile` where it builds the various buildboxes (6 places). I tested this out and ran a local `make buildbox-centos7-fips`, modified `Dockerfile-centos7` (the base docekrfile) and reran `make buildbox-centos7-fips` and it did the right thing, building a new fips image based on the modified Dockerfile.

Doing this with `--builder default` means it will work from the command line as well as CI.

I don't fully understand the `inline` cache, but I think it should be enabled with `--cache-to type=inline` instead of the buildkit arg as you can only cache to one cache, and this makes it clear you are using that one slot.

But overall it does seem to make sense to use the `docker` (default) builder, We only use the container builder for multi-arch builds of teleport itself. The buildboxes do not have multiarch builds (yet), so I support this change and can see no downside.

Commit:
swaps --build-arg BUILDKIT_INLINE_CACHE=1 for --cache-to type=inline

Commit:
adds DOCKER_BUILDER_NAME

ibeckermayer commented:
> I think instead of adding a parameter to the CI docker driver setup, we should just add `--builder default` in the `Makefile` where it builds the various buildboxes (6 places). I tested this out and ran a local `make buildbox-centos7-fips`, modified `Dockerfile-centos7` (the base docekrfile) and reran `make buildbox-centos7-fips` and it did the right thing, building a new fips image based on the modified Dockerfile.
> 
> Doing this with `--builder default` means it will work from the command line as well as CI.

Agreed that it would be ideal to specify this as a flag, however `default` will not always do the trick. For example in my local setup my builder with the docker driver is named `colima`. For that reason I added a variable to the Makefile in e2aca29d418efb84bee789ac83e351e15489a2d5 for figuring out which builder uses the `docker` driver.

> I don't fully understand the `inline` cache, but I think it should be enabled with `--cache-to type=inline` instead of the buildkit arg as you can only cache to one cache, and this makes it clear you are using that one slot.
> 
> But overall it does seem to make sense to use the `docker` (default) builder, We only use the container builder for multi-arch builds of teleport itself. The buildboxes do not have multiarch builds (yet), so I support this change and can see no downside.

I don't fully understand it either, but basically it means "put the cache alongside the image itself". The alternative options allow for caching to an external repository that's separate from the image. As I understand it, this becomes particularly relevant when doing multi-stage builds wherein you want to cache the intermediate stages without publishing the images themselves.

Confirmation that --cache-to type=inline does the same thing: https://docs.docker.com/build/cache/backends/inline/#synopsis. I changed it to use that since I agree its a better-integrated syntax.

Commit:
Revert "Release v16.0.0-dev.isaiah.18"

This reverts commit 6107dae6d8d17e43b950a67d092f4bbb254238a3.

camscale commented on build.assets/Makefile:
```diff
@@ -88,6 +88,35 @@ PULL_ON_CI = $(if $(filter $(CI),true),docker inspect --type=image $(1) >/dev/nu
+#   container-driver0 unix:///var/run/docker.sock inactive
+# default *           docker
+#   default           default                     running  20.10.23 linux/arm64, linux/arm/v7, linux/arm/v6
+DOCKER_BUILDER_NAME=$(shell docker buildx ls 2>/dev/null | awk '$$2 == "docker" || $$3 == "docker" {print $$1}')
```
nit: can you put spaces around the `=` - it's a little easier to read and Makefiles are tolerant of spaces there as opposed to shell scripts.

Also, you could hack the field separator in awk like this:

    awk -F '[ *]+' '$$2 == "docker" {print $$1}'

with a comment along the lines of "there may be an asterisk between the first and second column if that record is the default, so just subsume the asterisk into the field separator."

It is a shame that `docker buildx ls` is documented to take a `--format` arg, but does not actually take it (at least on my version of docker) - it would make this cleaner.

ibeckermayer replied:
```
awk -F '[ *]+' '$$2 == "docker" {print $$1}'
```

nice one!

> It is a shame that docker buildx ls is documented to take a --format arg, but does not actually take it (at least on my version of docker) - it would make this cleaner.

Yeah, mine neither. Docker is a mess as an overall project.

ibeckermayer replied:
9c41a6810d1eedbf1a9f170ba4c7e8ed6075c288

camscale commented on build.assets/Makefile:
```diff
@@ -88,6 +88,35 @@ PULL_ON_CI = $(if $(filter $(CI),true),docker inspect --type=image $(1) >/dev/nu
 # more details.
 RELEASE_TARGET ?= release
 
+# This variable extracts the name of the Docker buildx builder that uses the Docker driver.
```
Since this line is so far from the variable, perhaps word it as "DOCKER_BUILD_NAME extracts the name of ..." - that also matches Go style.

ibeckermayer replied:
9c41a6810d1eedbf1a9f170ba4c7e8ed6075c288

camscale commented on build.assets/Makefile:
```diff
@@ -189,10 +222,21 @@ buildbox-centos7:
 # Uses regular CentOS 7 buildbox as base, so limited arguments are needed.
 #
+# Because this target uses the image built by the buildbox-centos7 target, it is
+# expected to be run using a docker buildx driver. If another driver is used, the
```
s/a docker/the "docker"/

That currently reads like it just needs some docker buildx driver rather than the specific "docker" buildx driver.

ibeckermayer replied:
9c41a6810d1eedbf1a9f170ba4c7e8ed6075c288

camscale replied:
This is not address in that commit i think?

ibeckermayer replied:
As of that commit it says

> # DOCKER_BUILDER_NAME extracts the name of the buildx builder that uses the "docker" driver.

I think that addresses the original concern, but maybe I didn't understand it properly? Feel free to suggest something else.

camscale replied:
That's a different comment - this one says to replace 
> a docker buildx driver

with

> the "docker" buildx driver

camscale replied:
that is, in this sentence where my comment is attached, not in the heading.

ibeckermayer replied:
Oh my bad: 75b79ec889e8b02d468ef3bb6d17d37571ed2f9c

camscale commented on build.assets/Makefile:
```diff
@@ -204,12 +248,13 @@ buildbox-centos7-fips: buildbox-centos7
 buildbox-arm:
 	$(call PULL_ON_CI,$(BUILDBOX_ARM))
 	docker buildx build \
-		--build-arg BUILDBOX_VERSION=$(BUILDBOX_VERSION) \
```
Was this meant to be removed? It seems unrelated to the described changes.

ibeckermayer replied:
Ah yes, this was an opportunistic change since I noticed that Dockerfile-arm doesn't contain any such `ARG`

camscale reviewed (approved):
Minor non-impacting changes requested, but looks good otherwise. Lets give this a spin - this time it will work for sure!

Commit:
simplify DOCKER_BUILDER_NAME

ibeckermayer commented:
Taking her for another spin: https://drone.platform.teleport.sh/gravitational/teleport/33674

ibeckermayer commented:
Everything was broken on the last spin for an unrelated reason, here's another after a rebase: https://drone.platform.teleport.sh/gravitational/teleport/33691

Commit:
clarify comment wording

Commit:
updating e-ref to latest

Commit:
Merge branch 'master' into isaiah/docker-driver

codingllama commented:
FYI, this broke `make enter` for me:

codingllama commented:
(Edited previous comment with details)
